installer/modules.go

```diff
@@ -0,0 +1,935 @@
+		return NeedsRebootYes, nil
+	} else if output == "Automatic" {
+		log.Debug("Module needs host reboot")
+		return NeedsRebootAutomatic, nil
+	} else {
+		return NeedsRebootNo, fmt.Errorf("Unexpected reply from update module NeedsArtifactReboot query: %s",
+			output)
+	}
+}
+
+func (mod *ModuleInstaller) Reboot() error {
+	log.Debug("Executing ModuleInstaller.Reboot")
+	_, err := mod.callModule("ArtifactReboot", false)
+	return err
+}
+
+func (mod *ModuleInstaller) SupportsRollback() (bool, error) {
+	log.Debug("Executing ModuleInstaller.SupportsRollback")
+	output, err := mod.callModule("SupportsRollback", true)
+	if err != nil {
+		return false, err
+	} else if output == "" || output == "No" {
+		log.Debug("Module does not support rollback")
+		return false, nil
+	} else if output == "Yes" {
+		log.Debug("Module supports rollback")
+		return true, nil
+	} else {
+		return false, fmt.Errorf("Unexpected reply from update module SupportsRollback query: %s",
+			output)
+	}
+}
+func (mod *ModuleInstaller) RollbackReboot() error {
+	log.Debug("Executing ModuleInstaller.RollbackReboot")
+	_, err := mod.callModule("ArtifactRollbackReboot", false)
+	return err
+}
+
+func (mod *ModuleInstaller) CommitUpdate() error {
+	log.Debug("Executing ModuleInstaller.CommitUpdate")
+	_, err := mod.callModule("ArtifactCommit", false)
+	return err
+}
+
+func (mod *ModuleInstaller) Rollback() error {
+	log.Debug("Executing ModuleInstaller.Rollback")
+	_, err := mod.callModule("ArtifactRollback", false)
+	return err
+}
+
+func (mod *ModuleInstaller) VerifyReboot() error {
+	log.Debug("Executing ModuleInstaller.VerifyReboot")
+	_, err := mod.callModule("ArtifactVerifyReboot", false)
+	return err
+}
+
+func (mod *ModuleInstaller) VerifyRollbackReboot() error {
+	log.Debug("Executing ModuleInstaller.VerifyRollbackReboot")
+	_, err := mod.callModule("ArtifactVerifyRollbackReboot", false)
+	return err
+}
+
+func (mod *ModuleInstaller) Failure() error {
+	log.Debug("Executing ModuleInstaller.Failure")
+	_, err := mod.callModule("ArtifactFailure", false)
+	return err
+}
+
+func (mod *ModuleInstaller) Cleanup() error {
+	log.Debug("Executing ModuleInstaller.Cleanup")
+
+	payloadPath := mod.payloadPath()
+
+	// Prevent calling cleanup if the directory is already gone. Presumably
+	// it means that we already executed this, but spontaneously rebooted
+	// before we had time to finish everything. But if the tree is gone, it
+	// definitely executed the script first.
+	_, err := os.Stat(payloadPath)
+	if err != nil {
+		log.Infof("Could not access %s, assuming cleanup already done: %s",
+			payloadPath, err.Error())
+		return nil
+	}
+
+	_, modErr := mod.callModule("Cleanup", false)
+
+	err = os.RemoveAll(payloadPath)
+	if err != nil {
+		log.Errorf("Error during cleanup of module working directory: %s", err)
+	}
+
+	return modErr
+}
+
+func (mod *ModuleInstaller) GetType() string {
+	return mod.updateType
+}
+
+type ModuleInstallerFactory struct {
```

Comment:
I'm not 100% sure I understood what you meant, but I think some of the confusion here is that I called this struct `Factory`, but what it really implements is `UpdateStorerProducer`. I'll rename it to `ModuleInstallerProducer` perhaps. Since it sits right next to the `DualRootfs PayloadUpdatePerformerProducer` in the next struct, I think this should make the relationship clearer.